components/insight/SRC/org/openmicroscopy/shoola/env/data/views/calls/ThumbnailLoader.java

```diff
@@ -62,395 +59,305 @@
+        }
+        return Factory.createDefaultThumbnail("Error");
+    }
+
+    private ThumbnailStorePrx getThumbnailStore(PixelsData pxd) throws DSAccessException,
+            DSOutOfServiceException, ServerError {
+        // System.out.println(image.getId());
+        ThumbnailStorePrx store = service.createThumbnailStore(ctx);
+        if (!store.setPixelsId(pxd.getId())) {
+            store.resetDefaults();
+            store.setPixelsId(pxd.getId());
+        }
+        return store;
     }
 
     /**
-     * Creates a new instance.
-     * If bad arguments are passed, we throw a runtime exception so to fail
-     * early and in the caller's thread.
-     * 
-     * @param ctx The security context.
-     * @param pixelsID The id of the pixel set.
-     * @param maxWidth The m aximum acceptable width of the thumbnails.
-     * @param maxHeight The maximum acceptable height of the thumbnails.
-     * @param userID The user the thumbnail are for.
+     * Loads the thumbnail for {@link #images}<code>[index]</code>.
+     *
+     * @param pxd    The image the thumbnail for.
+     * @param userId The id of the user the thumbnail is for.
+     * @param store  The thumbnail store to use.
      */
-    public ThumbnailLoader(SecurityContext ctx, long pixelsID, int maxWidth,
-            int maxHeight, long userID)
-    {
-        if (maxWidth <= 0)
-            throw new IllegalArgumentException(
-                    "Non-positive id: "+pixelsID+".");
-        if (maxWidth <= 0)
-            throw new IllegalArgumentException(
-                    "Non-positive width: "+maxWidth+".");
-        if (maxHeight <= 0)
-            throw new IllegalArgumentException(
-                    "Non-positive height: "+maxHeight+".");
-        this.ctx = ctx;
-        pixelsCall = true;
-        this.maxWidth = maxWidth;
-        this.maxHeight = maxHeight;
-        this.pixelsID = pixelsID;
-        userIDs = new HashSet<Long>(1);
-        userIDs.add(userID);
-        service = context.getImageService();
+    private byte[] loadThumbnail(ThumbnailStorePrx store, PixelsData pxd, long userId)
+            throws ServerError, DSAccessException, DSOutOfServiceException {
+        int sizeX = maxWidth, sizeY = maxHeight;
+        if (asImage) {
+            sizeX = pxd.getSizeX();
+            sizeY = pxd.getSizeY();
+        } else {
+            Dimension d = Factory.computeThumbnailSize(sizeX, sizeY,
+                    pxd.getSizeX(), pxd.getSizeY());
+            sizeX = d.width;
+            sizeY = d.height;
+        }
+
+        if (userId >= 0) {
+            long rndDefId = service.getRenderingDef(ctx,
+                    pxd.getId(), userId);
+            // the user might not have own rendering settings
+            // for this image
+            if (rndDefId >= 0)
+                store.setRenderingDefId(rndDefId);
+        }
+
+        if (VersionCompare.compare(context.getGateway().getServerVersion(), VERSION_THUMBNAIL_NO_DEFAULT) >= 0) {
```

Comment:
@dominikl and @jburel will know better, but I assume there is a location that a `boolean server_5_4_8_or_later` could be saved so as to not need this in the loop.

Reply:
It's not a big issue as the gateway holds the server version as string, it doesn't request it over and over again. But if you want to keep it in Insight directly, you could add a `LookupNames.SERVER_VERSION` and 'bind' it to the Registry when Insight connects to the server somewhere around here: https://github.com/openmicroscopy/openmicroscopy/blob/ffa5419fb669871f959973f68b2c3fa0e4e2b576/components/insight/SRC/org/openmicroscopy/shoola/env/data/DataServicesFactory.java#L575 .

Reply:
Here's where Gateway obtains it's server version during connection https://github.com/openmicroscopy/openmicroscopy/blob/040837bf61ce493d0a742a3fe871acc60847407d/components/blitz/src/omero/gateway/Gateway.java#L1066

Reply:
I'll leave it to you guys to work out what's most "Insight-esque";  my point is mostly about whether or not the albeit slight overhead of a redudant check for each and every thumbnail is a concern.